drivers/spi/spi_shell.c

```diff
@@ -4,54 +4,278 @@
+
+/* Instantiate spi_shell_device struct from node_id, if node_id is a spi device. */
+#define INST_ALL_SPI_DEVICES_AS_SPI_SHELL_DEVICES(node_id)                                         \
+	RUN_FN_ON_SPI_DEVICE(node_id, INST_SPI_SHELL_DEVICE_AS_SPI_DEV(node_id))
+
+/* List of spi shell devices. At compile time we instantiate structs for all spi devices.
+ * Additional empty space is reserved for the spi buses, which are added in spi_buses_init at
+ * runtime.
+ */
+static struct spi_shell_device spi_shell_devices[] = {
+	DT_FOREACH_STATUS_OKAY_NODE(INST_ALL_SPI_DEVICES_AS_SPI_SHELL_DEVICES)
+		CREATE_NUM_EMPTY_STRUCTS(CONFIG_SPI_SHELL_NUM_EMPTY_INSTANCES)};
+static size_t num_spi_shell_devices =
+	ARRAY_SIZE(spi_shell_devices) - CONFIG_SPI_SHELL_NUM_EMPTY_INSTANCES;
+
+#define INST_MAPS_FROM_SPI_DEVICE_NODELABELS(node_id)                                              \
+	RUN_FN_ON_SPI_DEVICE(node_id, DT_FOREACH_NODELABEL_VARGS(node_id, INST_MAP, node_id))
+
+#define INST_MAPS_FROM_SPI_DEVICE_NODE_ID(node_id)                                                 \
+	RUN_FN_ON_SPI_DEVICE(node_id, INST_MAP_FROM_NODE_ID(node_id))
+
+/* A list of maps. At compile time we create maps for all nodelabels and node_ids of spi devices.
+ * Additional empty space is reserved for the spi buses, which are added in spi_buses_init at
+ * runtime.
+ */
+static struct map maps[] = {
+	DT_FOREACH_STATUS_OKAY_NODE(INST_MAPS_FROM_SPI_DEVICE_NODELABELS)
+		DT_FOREACH_STATUS_OKAY_NODE(INST_MAPS_FROM_SPI_DEVICE_NODE_ID)
+			CREATE_NUM_EMPTY_STRUCTS(CONFIG_SPI_SHELL_NUM_EMPTY_INSTANCES)};
+static size_t num_maps = ARRAY_SIZE(maps) - CONFIG_SPI_SHELL_NUM_EMPTY_INSTANCES;
 
 static bool device_is_spi(const struct device *dev)
 {
 	return DEVICE_API_IS(spi, dev);
 }
 
-static void device_name_get(size_t idx, struct shell_static_entry *entry)
+static bool device_is_gpio(const struct device *dev)
+{
+	return DEVICE_API_IS(gpio, dev);
+}
+
+/**
+ * @brief Initialize spi buses at runtime.
+ *
+ * Since Zephyr currently doesn't support getting a device for all spi buses at compile time in a
+ * generic way, we do it at runtime.
+ *
+ * For each spi bus device we:
+ * - add an entry to spi_shell_devices array
+ * - add an entry to maps array by its name
+ * - add an entry to maps array for each it's nodelabel
+ */
+static int spi_buses_init(void)
+{
+	int idx = 0;
+
+	while (1) {
+		const struct device *dev = shell_device_filter(idx, device_is_spi);
+
+		idx++;
+
+		if (dev == NULL) {
+			break;
+		}
+
+		if (num_spi_shell_devices == ARRAY_SIZE(spi_shell_devices)) {
+			printk("ERROR: not enough space in spi_shell_devices array\n");
+			printk("Increase CONFIG_SPI_SHELL_NUM_EMPTY_INSTANCES.\n");
+			break;
+		}
+
+		spi_shell_devices[num_spi_shell_devices++] =
+			INST_SPI_SHELL_DEVICE_AS_SPI_DEV_AS_SPI_BUS(dev);
+
+		maps[num_maps++] = (struct map){
+			.label = dev->name,
+			.name = dev->name,
+		};
+
+		const struct device_dt_nodelabels *nl = device_get_dt_nodelabels(dev);
```

Comment:
This means CONFIG_DEVICE_DT_METADATA is enabled, so make it depending on it.
(it takes ROM and not all users have ROM to spare)